### PR TITLE
[Clang][Driver] Warn about `-c/-S` with `-fsyntax-only`

### PR DESCRIPTION
**Pull request by chestnykh**
Emit warning that `-S` and/or `-c` arguments are not used if `-fsyntax-only` is also passed to clang
`addPGOAndCoverageFlags` is not the right place
to produce this warning
Now `-fsyntax-only -c/-S` combination handles like `-fsyntax-only -E` in `BuildJobs()` driver function

### CODE REVIEW

**Comment by chestnykh** [comment]
CC: @shafik @ChuanqiXu9

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -794,10 +794,6 @@ static void addPGOAndCoverageFlags(const ToolChain &TC, Compilation &C,
       Args.hasArg(options::OPT_coverage))
     FProfileDir = Args.getLastArg(options::OPT_fprofile_dir);
 
-  // TODO: Don't claim -c/-S to warn about -fsyntax-only -c/-S, -E -c/-S,
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/Clang.cpp:797]: Thanks for catching this. I left this in a07b135ce0c0111bd83450b5dc29ef0381cdbc39 to be cleaned up later but I forgot to remove this TODO.

File: clang/test/Driver/warn-fsyntax-only-c-S.c [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang -fsyntax-only -S %s 2>&1 | FileCheck %s --check-prefix=CHECK-ASM
```
Comment (MaskRay) [inline clang/test/Driver/warn-fsyntax-only-c-S.c:1]: Perhaps just warn-fsyntax-only.c

File: clang/test/Driver/warn-fsyntax-only-c-S.c [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+// RUN: %clang -fsyntax-only -c %s 2>&1 | FileCheck %s --check-prefix=CHECK-OBJ
+// RUN: %clang -fsyntax-only -S -c %s 2>&1 | FileCheck %s --check-prefix=CHECK-BOTH
+
+// CHECK-ASM: warning: argument unused during compilation: '-S' [-Wunused-command-line-argument]
+// CHECK-OBJ: warning: argument unused during compilation: '-c' [-Wunused-command-line-argument]
+
+// CHECK-BOTH: warning: argument unused during compilation: '-S' [-Wunused-command-line-argument]
+// CHECK-NEXT: warning: argument unused during compilation: '-c' [-Wunused-command-line-argument]
```
Comment (MaskRay) [inline clang/test/Driver/warn-fsyntax-only-c-S.c:9]: CHECK-NEXT is not used by any FileCheck command

File: clang/test/Driver/warn-fsyntax-only-c-S.c (hunk 2 above)
Comment (chestnykh) [reply clang/test/Driver/warn-fsyntax-only-c-S.c]: Renamed

File: clang/test/Driver/warn-fsyntax-only-c-S.c (hunk 3 above)
Comment (chestnykh) [reply clang/test/Driver/warn-fsyntax-only-c-S.c]: I've added `-fsyntax-only -E` case. `-c -S` i think is like duplication of `-c` and `-S` test cases

**Review by MaskRay — APPROVED** [review]

**Comment by dyung** [comment]
@chestnykh Can you take a look at the bot failures and revert if you need time to investigate so we can get the bots back to green?

**Comment by mstorsjo** [comment]
FWIW, in addition to `-c` in combination with `-fsyntax-only`, this now also warns for `-c` in combination with `-E`.

That sounds like a reasonable thing to do to me, so it's probably fine, but it'd be nice to have that aspect acknowledged as one intended behaviour change as well.

**Comment by MaskRay** [comment]
> FWIW, in addition to `-c` in combination with `-fsyntax-only`, this now also warns for `-c` in combination with `-E`.
> 
> That sounds like a reasonable thing to do to me, so it's probably fine, but it'd be nice to have that aspect acknowledged as one intended behaviour change as well.

Thanks for pointing this out. Yes, the warning for `-c -E` seems desired. It's a bit difficult to analyze what combinations lead to warnings due to our `claim` code scattering all over the driver....

**Pull request merged**